lib/logstash/filters/grok/timeout_enforcer.rb

```diff
@@ -64,44 +68,32 @@ def stop!
 
   def start_thread_groking(thread)
     # Clear any interrupts from any previous invocations that were not caught by Joni
-    thread.interrupted
-    synchronize do
-      @threads_to_start_time[thread] = java.lang.System.nanoTime()
-    end
+    java.lang.Thread.interrupted
+    @threads_to_start_time.put(thread, java.lang.System.nanoTime)
   end
 
   def stop_thread_groking(thread)
-    synchronize do
-      @threads_to_start_time.delete(thread)
-    end
+    @threads_to_start_time.remove(thread)
   end
 
   def cancel_timed_out!
-    synchronize do
-      @threads_to_start_time.each do |thread,start_time|
-        now = java.lang.System.nanoTime # save ourselves some nanotime calls
-        elapsed = java.lang.System.nanoTime - start_time
-        if elapsed > @timeout_nanos
-          elapsed_millis = elapsed / 1000
-          thread.interrupt()
-          # Ensure that we never attempt to cancel this thread twice in the event
-          # of weird races
-          stop_thread_groking(thread)
+    now = java.lang.System.nanoTime # save ourselves some nanotime calls
+    @threads_to_start_time.entry_set.each do |entry|
+      start_time = entry.get_value
+      if start_time < now && now - start_time > @timeout_nanos
+        thread  = entry.get_key
+        # Ensure that we never attempt to cancel this thread twice in the event
+        # of weird races
+        if stop_thread_groking(thread)
```

Comment:
I still don't see how we know that at this point `thread` is working on the same task as on line 83 where we have checked the time elapsed. Seems like a race to me, am I missing something?

Reply:
@andrewvc you're not, but I think this is a risk we can cleanly take, because with the latest changes we will never interrupt a thread that isn't grokking (sorry for failing to comment on this more explicitly).
Think about the orders of magnitude here:

* The cancel thread will do an iteration over a very small number of entries, some numeric comparisons and (if it indeed does have to interrupt some stuff) a `volatile` write.

In order to kill the wrong grok action, the the filter + output + poll + from queue + start new grokking action in the worker would have to be faster than a numeric comparison + `ConcurrentHashMap#delete` here.
I think that's close to theoretically impossible and practically impossible. I can see how you could make an argument for this happening by virtue of some freak scheduling decision in the kernel, but practically it's impossible imo. Even if there is a theoretical chance (and this is easily less than 1:1M imo looking at the number of cycles the worker has to take for this to happen vs. the number of cycles the enforcer thread uses up), the effect is at worst wrongfully cancelling a grok action.

I can add a sequence number to this to make it 100% safe (that was my initial plan I explained above), but imo it wasn't worth it performance (small hit, but a hit) + complexity wise ( another 10 lines).